patches/0010-add-appinsights-telemetry.patch

```diff
@@ -0,0 +1,1746 @@
++		if addr == "" {
++			addr = "localhost:0"
++		}
++		l, err := net.Listen("tcp", addr)
++		if err != nil {
++			log.Fatal(err)
++		}
++		*appInsightsAddr = l.Addr().String()
++		appInsightsURL = "http://" + *appInsightsAddr
++		fmt.Fprintf(os.Stderr, "go test App Insights server running at MS_GOTOOLCHAIN_TELEMETRY_ENDPOINT=%s\n", appInsightsURL)
++		go func() {
++			log.Fatalf("App Insights: http.Serve: %v", http.Serve(l, http.HandlerFunc(appInsightsHandler)))
++		}()
++	})
++}
++
++func appInsightsHandler(w http.ResponseWriter, r *http.Request) {
++	appInsightsRequests.Add(1)
++	fmt.Fprint(w, `{"itemsReceived": 1, "itemsAccepted": 1}`)
++}
+diff --git a/src/cmd/go/script_test.go b/src/cmd/go/script_test.go
+index 1345ea8bb8e530..d3f50e5c24e6c9 100644
+--- a/src/cmd/go/script_test.go
++++ b/src/cmd/go/script_test.go
+@@ -60,6 +60,7 @@ func TestScript(t *testing.T) {
+ 	}
+ 
+ 	StartProxy()
++	StartAppInsightsServer()
+ 
+ 	var (
+ 		ctx         = context.Background()
+@@ -254,6 +255,9 @@ func scriptEnv(srv *vcstest.Server, srvCertFile string) ([]string, error) {
+ 		"HGRCPATH=",
+ 		"GOTOOLCHAIN=auto",
+ 		"MS_GOTOOLCHAIN_ALLOW_NON_LOCAL=1", // allow non-local toolchains, some tests expect GOTOOLCHAIN to be honored
++		"MS_GOTOOLCHAIN_TELEMETRY_ENDPOINT=" + appInsightsURL,
++		"MS_GOTOOLCHAIN_TELEMETRY_ALLOW_GO_DEVEL=on", // allow telemetry for Go development versions
++		"MS_GOTOOLCHAIN_TELEMETRY_INSTRUMENTATION_KEY=fake_key",
+ 		"newline=\n",
+ 	}
+ 
+@@ -404,6 +408,9 @@ func readCounters(t *testing.T, telemetryDir string) map[string]uint64 {
+ func checkCounters(t *testing.T, telemetryDir string) {
+ 	counters := readCounters(t, telemetryDir)
+ 	if _, ok := scriptGoInvoked.Load(testing.TB(t)); ok {
++		if appInsightsRequests.Load() == 0 {
++			t.Fatal("go was invoked but no App Insights request were recorded")
++		}
+ 		if !disabledOnPlatform && len(counters) == 0 {
+ 			t.Fatal("go was invoked but no counters were incremented")
+ 		}
+diff --git a/src/cmd/internal/telemetry/counter/counter.go b/src/cmd/internal/telemetry/counter/counter.go
+index 5cef0b0041551a..64c3ac0b74fc15 100644
+--- a/src/cmd/internal/telemetry/counter/counter.go
++++ b/src/cmd/internal/telemetry/counter/counter.go
+@@ -7,12 +7,20 @@
+ package counter
+ 
+ import (
++	"context"
+ 	"flag"
++	"log"
+ 	"os"
++	"time"
+ 
+ 	"golang.org/x/telemetry/counter"
++
++	mstelemetry "github.com/microsoft/go-infra/telemetry"
++	mscounter "github.com/microsoft/go-infra/telemetry/counter"
+ )
+ 
++var appInsightsInstrumentationKey = os.Getenv("MS_GOTOOLCHAIN_TELEMETRY_INSTRUMENTATION_KEY") // TODO: hardcode to the correct value once we have a real key
++
+ var openCalled bool
+ 
+ func OpenCalled() bool { return openCalled }
+@@ -22,11 +30,29 @@ func OpenCalled() bool { return openCalled }
+ func Open() {
+ 	openCalled = true
+ 	counter.OpenDir(os.Getenv("TEST_TELEMETRY_DIR"))
++	if appInsightsInstrumentationKey != "" && os.Getenv("MS_GOTOOLCHAIN_TELEMETRY_ENABLED") != "off" {
```

Comment:
I think we should accept 1 and 0, not (only) words, to avoid mistype/misread/misremember. Words should also not be case sensitive if we support them.

Reply:
Lets go for the 1 and 0 instead of words 👍 .